### PR TITLE
Ensuring ServiceBusReactorReceiver is closed when the link is disposed

### LINKED ISSUE #26528: Object leak in ServiceBus acceptNextSession() API
The following is a minimum repro extracted from a user's large project.

```java
public class App {
    public static void main(String[] args) {
        ServiceBusClientBuilder sbClientBuilder = new ServiceBusClientBuilder()
                .connectionString(System.getenv("CON_STR"))
                .retryOptions(new AmqpRetryOptions()
                        .setMode(AmqpRetryMode.FIXED)
                        .setTryTimeout(Duration.ofSeconds(30))
                        .setDelay(Duration.ofSeconds(1))
                        .setMaxDelay(Duration.ofSeconds(5))
                        .setMaxRetries(0));
        
        ServiceBusSessionReceiverClientBuilder sbSessionReceiverClientBuilder = sbClientBuilder
                .sessionReceiver()
                .maxAutoLockRenewDuration(Duration.ofMinutes(0))
                .queueName(System.getenv("SESSION_QUEUE_NAME"))
                .receiveMode(ServiceBusReceiveMode.PEEK_LOCK)
                .prefetchCount(0);
        try (ServiceBusSessionReceiverClient sbSessionReceiverClient = sbSessionReceiverClientBuilder.buildClient()) { 
            // Accept 20 sessions.            
            for (int i = 0; i < 20; i++) {                
                try {                    
                    System.out.println("Accepting session #" + i);                    
                    try (ServiceBusReceiverClient receiverClient = sbSessionReceiverClient.acceptNextSession()) {    
                        
                    }                
                } catch (IllegalStateException ise) {  
                    // ise is thrown if service identifies that there is no session to return.     
                    System.out.println(ise.getMessage());         
                } catch (Throwable t) {       
                    t.printStackTrace();         
                }          
            }      
        }    
    }
}
```

If I run this code on a session-enabled queue "without any message", each acceptNextSession call will eventually timeout because there is no session to connect. Each time the API timeout, an object of `ServiceBusReactorReceiver` is leaked.

ServiceBusReactorReceiverLeak

### PR DESCRIPTION
**Pull request by anuchandy**
See the git_ticket https://github.com/Azure/azure-sdk-for-java/issues/26528 for details.

Fixes #26528

### CODE REVIEW

File: eng/versioning/version_client.txt [hunk 1]
```diff
@@ -66,6 +66,7 @@ com.azure:azure-containers-containerregistry;1.0.0;1.1.0-beta.1
 com.azure:azure-containers-containerregistry-perf;1.0.0-beta.1;1.0.0-beta.1
 com.azure:azure-core;1.24.1;1.25.0-beta.1
 com.azure:azure-core-amqp;2.3.7;2.4.0-beta.1
+unreleased_com.azure:azure-core-amqp;2.4.0-beta.1
```
Comment (alzimmermsft) [inline eng/versioning/version_client.txt:69]: nit: put this in the unreleased section near the bottom of this file

File: sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/ReactorReceiver.java [hunk 2]
```diff
@@ -268,13 +270,18 @@ protected Message decodeDelivery(Delivery delivery) {
 
     /**
      * Disposes of the receiver when an exception is encountered.
+     * <br/>
```
Comment (alzimmermsft) [inline sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/ReactorReceiver.java:273]: nit: use a new paragraph tag

File: sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/ReactorReceiver.java [hunk 3]
```diff
@@ -268,13 +270,18 @@ protected Message decodeDelivery(Delivery delivery) {
...
+     * While {@link ReactorReceiver#closeAsync()} exposes disposal API through {@link AsyncCloseable}
+     * contract, this closeAsync(String, ErrorCondition) perform the same disposal, but with additional
+     * contextual information, example for such a context are if resource needs to be disposed of
```
Comment (alzimmermsft) [inline sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/ReactorReceiver.java:276]:
```suggestion
     * contextual information. For example, the context may indicate if the resource needs to be disposed of
```

**Review by conniey — APPROVED** [review]
Should we also update Event Hubs to use this unreleased_ tag to make sure nothing breaks in between?

**Review by alzimmermsft — APPROVED** [review]

**Pull request merged**